replica/table_sstable_set.cc

```diff
@@ -0,0 +1,240 @@
+        return set->for_each_sstable_until(func);
+    });
+}
+
+future<stop_iteration> table_sstable_set::for_each_sstable_gently_until(std::function<future<stop_iteration>(const shared_sstable&)> func) const {
+    return for_each_sstable_set_gently_until(query::full_partition_range, [func = std::move(func)] (lw_shared_ptr<sstable_set> set) {
+        return set->for_each_sstable_gently_until(func);
+    });
+}
+
+bool table_sstable_set::insert(shared_sstable sst) {
+    throw_with_backtrace<std::bad_function_call>();
+}
+bool table_sstable_set::erase(shared_sstable sst) {
+    throw_with_backtrace<std::bad_function_call>();
+}
+
+size_t
+table_sstable_set::size() const noexcept {
+    size_t ret = 0;
+    for_each_sstable_set_until(query::full_partition_range, [&] (lw_shared_ptr<sstable_set> set) {
+        ret += set->size();
+        return stop_iteration::no;
+    });
+    return ret;
+}
+
+uint64_t
+table_sstable_set::bytes_on_disk() const noexcept {
+    uint64_t ret = 0;
+    for_each_sstable_set_until(query::full_partition_range, [&] (lw_shared_ptr<sstable_set> set) {
+        ret += set->bytes_on_disk();
+        return stop_iteration::no;
+    });
+    return ret;
+}
+
+class table_incremental_selector : public incremental_selector_impl {
+    table& _table;
+
+    // _cur_set and _cur_selector contain a snapshot
+    // for the currently selected compaction_group.
+    // _lowest_next_position corresponds the start position
+    // of the compaction_group following the currently selected one
+    lw_shared_ptr<sstable_set> _cur_set;
+    std::optional<sstable_set::incremental_selector> _cur_selector;
+    dht::ring_position_ext _lowest_next_position = dht::ring_position_ext::max();
+
+public:
+    table_incremental_selector(table& t)
+            : _table(t)
+    {
+        if (auto cg = _table.single_compaction_group_if_available()) {
+            _cur_set = cg->make_compound_sstable_set();
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+        }
+    }
+
+    virtual std::tuple<dht::partition_range, std::vector<shared_sstable>, dht::ring_position_ext> select(const dht::ring_position_view& pos) override {
+        // Always return minimum singular range, such that incremental_selector::select() will always call this function,
+        // which in turn will find the next sstable set to select sstables from.
+        const dht::partition_range current_range = dht::partition_range::make_singular(dht::ring_position::min());
+
+        if (!_cur_selector) {
+            auto* sg = _table.storage_group_for_token(pos.token());
+            _cur_set = sg->make_compound_sstable_set();
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+            auto next_token = sg->main_compaction_group()->next_compaction_group_token();
+            _lowest_next_position = next_token.is_maximum()
```

Comment:
I think it makes more sense to move next_compaction_group_token() impl to storage_group. Ideally, we retrieve this info from table itself.

It could happen that the adjacent tablet range is not present in this shard, so we'd come back in select() needlessly. It's better for table to return the next *available* token, based on presence of tablets in this shard. Today, table can feed the next token of adjacent group, but it will change once I implement https://github.com/scylladb/scylladb/issues/16877

Reply:
For that, you probably will have to work with ids.

size_t storage_group_idx_for_token(dht::token);
const storage_group* storage_group_at(size_t);

then you retrieve both current and next, and retrieve token range from next, to calculate next token.

Reply:
ok, good idea